pkg/storegateway/series_refs.go

```diff
@@ -1030,6 +1034,52 @@ func (s *loadingSeriesChunkRefsSetIterator) loadSeries(ref storage.SeriesRef, lo
 	return lset, s.chunkMetasBuffer, nil
 }
 
+type filteringSeriesChunkRefsSetIterator struct {
+	from     seriesChunkRefsSetIterator
+	matchers []*labels.Matcher
+
+	current seriesChunkRefsSet
+}
+
+func (m *filteringSeriesChunkRefsSetIterator) Next() bool {
+	if !m.from.Next() {
+		return false
+	}
+
+	next := m.from.At()
+	writeIdx := 0
+
+	for _, series := range next.series {
+		matches := true
+		for _, matcher := range m.matchers {
+			if !matcher.Matches(series.lset.Get(matcher.Name)) {
```

Comment:
I wonder if this can be optimized somehow :-/
I'd expect Prometheus to have a `matchers.Match(labels.Labels)` code, to be used everywhere.

Reply:
Map lookup is typically slower than iterating a slice over and over, if there are few elements inside (we've seen it also when working on the regexp matcher optimizations).

I think one way to optimise this is using `labels.Labels.Range()` and iterate the series labels only once. We could sort the matchers by `Name`, then use `Range()` to iterate labels sorted by name (labels are sorted by name by contract) and advance the index of the "matcher to check" accordingly (correctly handling the case we're skipping a matcher because no matching label has been found).

However, I wouldn't block on this and keep it _eventually_ for a follow up PR if we see that this code path really impacts performance.

Reply:
sounds good to me